router/find_path.gno

```diff
@@ -0,0 +1,115 @@
+) (swapPaths SwapPaths) {
+	tokenPairs := TokenPairs{}
+	poolList := p.GetPoolList()
+
+	for i, poolPath := range poolList {
+		token0Path, token1Path, pFee := poolPathWithFeeDivide(poolPath)
+
+		{
+			k := token0Path
+			v := token1Path + ":" + strconv.Itoa(pFee)
+
+			tokenPairs[k] = append(tokenPairs[k], v)
+		}
+
+		{
+			k := token1Path
+			v := token0Path + ":" + strconv.Itoa(pFee)
+
+			tokenPairs[k] = append(tokenPairs[k], v)
+		}
+	}
+
+	swapPaths = getSwapPaths(tokenPairs, inputTokenPath, outputTokenPath)
+	return swapPaths
+}
+
+func getSwapPaths(
+	tokenPairs TokenPairs,
+	inputTokenPath string,
+	outputTokenPath string,
+) (swapPaths SwapPaths) {
+	swapPaths = make(SwapPaths, 0)
+
+	// check if there is path that starts with input
+	require(len(tokenPairs[inputTokenPath]) != 0, ufmt.Sprintf("[ROUTER] find_path.gno__getSwapPaths() || len(tokenPairs[inputTokenPath]) == 0, inputTokenPath: %s", inputTokenPath))
+
+	// find direct path
+	for _, output := range tokenPairs[inputTokenPath] {
+		if strings.HasPrefix(output, outputTokenPath) {
+			outputPath, outputFee := singlePoolPathWithFeeDivide(output)
+			directPath := inputTokenPath + "," + outputFee + "," + outputPath
+			swapPaths[len(swapPaths)] = directPath
+
+			tokenPairs[inputTokenPath] = removeItemFromStringArray(tokenPairs[inputTokenPath], output)
+		}
+	}
+
+	// find nested path
+	swapPaths = findTwicePath(tokenPairs, inputTokenPath, outputTokenPath, swapPaths)
+	swapPaths = findThreeTimePath(tokenPairs, inputTokenPath, outputTokenPath, swapPaths)
+
+	return swapPaths
+}
+
+func findTwicePath(
+	tokenPairs TokenPairs,
+	inputTokenPath string,
+	outputTokenPath string,
+	swapPaths SwapPaths,
+) SwapPaths {
+	for _, second := range tokenPairs[inputTokenPath] {
+		secondPath, secondFee := singlePoolPathWithFeeDivide(second)
+
+		for _, third := range tokenPairs[secondPath] {
+			thirdPath, thirdFee := singlePoolPathWithFeeDivide(third)
+
+			if strings.HasPrefix(third, outputTokenPath) {
+				nestedPath := inputTokenPath + "," + secondFee + "," + secondPath + "," + thirdFee + "," + outputTokenPath
+				swapPaths[len(swapPaths)] = nestedPath
+			}
+		}
+	}
+
+	return swapPaths
+}
+
+func findThreeTimePath(
+	tokenPairs TokenPairs,
+	inputTokenPath string,
+	outputTokenPath string,
+	swapPaths SwapPaths,
+) SwapPaths {
+	for _, second := range tokenPairs[inputTokenPath] {
+		secondPath, secondFee := singlePoolPathWithFeeDivide(second)
+
+		for _, third := range tokenPairs[secondPath] { // bar > bz
+			thirdPath, thirdFee := singlePoolPathWithFeeDivide(third)
+
+			for _, fourth := range tokenPairs[thirdPath] {
+				fourthPath, fourthFee := singlePoolPathWithFeeDivide(fourth)
+
+				if strings.HasPrefix(fourth, outputTokenPath) {
+					// three times
+					nestedPath := inputTokenPath + "," + secondFee + "," + secondPath + "," + thirdFee + "," + thirdPath + "," + fourthFee + "," + fourthPath
+					swapPaths[len(swapPaths)] = nestedPath
+				}
+			}
+		}
+	}
+	return swapPaths
+}
```

Comment:
I think we can combine `findTwicePath` and `findThreeTimesPath` into one. There are two ways to do this off the top of my head. One is the parameterize how many times to traverse and then "run it as a recursive function", and the other is to use a "graph algorithm".

The first one is easy to fix and does not involve many changes. For example, we could change like this

```go
func findPath(
    tokenPairs TokenPairs,
    currentTokenPath string,
    outputTokenPath string,
    currentPath string,
    remainingHops int,
    swapPaths *SwapPaths,
) {
    if remainingHops == 0 {
        if strings.HasPrefix(currentTokenPath, outputTokenPath) {
            *swapPaths = append(*swapPaths, currentPath)
        }
        return
    }

    for _, next := range tokenPairs[currentTokenPath] {
        nextPath, nextFee := singlePoolPathWithFeeDivide(next)
        newPath := currentPath
        if currentPath != "" {
            newPath += ","
        }
        newPath += nextFee + "," + nextPath
        findPath(tokenPairs, nextPath, outputTokenPath, newPath, remainingHops-1, swapPaths)
    }
}
```
However, there is a potential for overflow, so you'll need to set a separate maximum value. I prefer the second method because it doesn't improve time complexity much.

You can think of a graph with token pairs as edges and tokens as nodes. Also, if I understand correctly, the fee information in the pool can be used as a weight.

If you use the pool as a weight information, we can use an algorithm like [Dijkstra](https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm) or [Bellman-Ford](https://en.wikipedia.org/wiki/Bellman%E2%80%93Ford_algorithm), but I think that simply using an algorithm like BFS can significantly reduce the time complexity from $O(n^2)$ to $O(V + E)$.

However, implementing a graph algorithm is quite cumbersome, so if you don't mind, you can modify it later.

Reply:
applied cb16f6a9b977b1bead8de7855f4f819141ee85cd